lib/crypto/aes.toit

```diff
@@ -2,92 +2,155 @@
 // Use of this source code is governed by an MIT-style license that can be
 // found in the lib/LICENSE file.
 
+/**
+Base class for the hardware accelerated Advanced Encryption Standard (AES).
+
+See https://en.wikipedia.org/wiki/Advanced_Encryption_Standard.
+
+AES has multiple different modes it can use.
+The ones implemented here are CBC and ECB.
+*/
+abstract class Aes:
+  aes_ := ?
+
+  /**
+  Initialize an Aes class from a subclass. 
+  If the $initialization_vector is empty, then AES ECB mode is selected.
+  If the $initialization_vector has length 16, then AES CBC mode is selected.
+  */
+  constructor.from_subclass key/ByteArray initialization_vector/ByteArray encrypt/bool:
+    aes_ = aes_init_ resource_freeing_module_ key initialization_vector encrypt
+    add_finalizer this:: this.close
+
+  /**
+  Encrypts the given $cleartext.
+
+  The size of the $cleartext must be a multiple of 16.
+  Returns a byte array with the encrypted data.
+  */
+  encrypt cleartext/ByteArray -> ByteArray:
+    return crypt_ cleartext true
+
+  /**
+  Decrypts the given $ciphertext.
+  The size of the $ciphertext must be a multiple of 16.
+  Returns a byte array with the decrypted data.
+  */
+  decrypt ciphertext/ByteArray -> ByteArray:
+    if not aes_: throw "ALREADY_CLOSED"
+    return crypt_ ciphertext false
+
+  /**
+  Calls the assosiated primitive for the selected AES mode.
+  */
+  abstract crypt_ input/ByteArray encrypt/bool -> ByteArray
```

Comment:
Maybe encrypt should be a named argument? Makes the call sites a little clearer, but given that this is a private method, I don't think it is super important.

```
crypt_ ciphertext --no-encrypt
```

Reply:
Switched the encrypt bool to a named argument for the crypt_ method and for the initialize_ constructor.